### PR TITLE
[Tesla] Add SSO handler to authenticate against Tesla SSO service

### LINKED ISSUE #10016: [Tesla] Authentication no longer working / getting blocked
Since Tesla changed their API, authentication is no longer working. Current binding, using a new refresh token yields an HTTP error 400. Using Postman with the URL being used with this binding, I do get an HTTP/400, with JSON return data, "response" set to "endpoint_deprecated:_please_update_your_app.".

The changes that Tesla has done, has been documented here: https://github.com/timdorr/tesla-api/issues/260

The new authentication procedure has been documented here: https://tesla-api.timdorr.com/api-basics/authentication
(As always, a big "Thank you!!" to Tim Dorr!

## Expected Behavior
Tesla account comes online as soon as you enter the refresh_token and save configuration

## Current Behavior
Account does not come online, but reports "Communication error", and "HTTP Code 400" or "HTTP Code 444".

## Possible Solution
Replace authentication procedure and current URLs in code.

## Steps to Reproduce (for Bugs)
Refresh Tesla token configuration in OpenHAB.

### PR DESCRIPTION
**Pull request by cguedel**
Signed-off-by: Christian Güdel <cg@dmesg.ch>

Fixes #10016

Adds authentication against the Tesla SSO service as described here https://tesla-api.timdorr.com/api-basics/authentication

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Looks like a productive weekend, thanks!
Will try to (start to) test & review tonight.

**Comment by kaikreuzer** [comment]
First feedback: Creating an account thing (without filling in credentials or token) in the UI immediately results in two errors in the log:
```
17:26:27.024 [-thingHandler-1] ERROR o.o.b.t.i.handler.TeslaSSOHandler:253 - An exception occurred while invoking a HTTP request: 'java.lang.IllegalArgumentException: Buffering capacity 16384 exceeded'
17:26:27.024 [-thingHandler-1] ERROR o.o.b.t.i.handler.TeslaSSOHandler:178 - Failed to obtain code from SSO login page when submitting form, response status code: no response
```

**Comment by kaikreuzer** [comment]
Second feedback: Vehicle discovery crashes with an exception:
```
17:33:47.439 [-thingHandler-3] WARN  o.o.c.i.c.WrappedScheduledExecutorService:67 - Scheduled runnable ended with an exception: 
javax.ws.rs.ProcessingException: java.io.IOException: IOException invoking https://owner-api.teslamotors.com/api/1/vehicles: HTTPS hostname wrong:  should be <owner-api.teslamotors.com>
	at org.apache.cxf.jaxrs.client.AbstractClient.checkClientException(AbstractClient.java:633)
	at org.apache.cxf.jaxrs.client.AbstractClient.preProcessResult(AbstractClient.java:607)
	at org.apache.cxf.jaxrs.client.WebClient.doResponse(WebClient.java:1145)
	at org.apache.cxf.jaxrs.client.WebClient.doChainedInvocation(WebClient.java:1082)
	at org.apache.cxf.jaxrs.client.WebClient.doInvoke(WebClient.java:927)
	at org.apache.cxf.jaxrs.client.WebClient.doInvoke(WebClient.java:896)
	at org.apache.cxf.jaxrs.client.WebClient.invoke(WebClient.java:461)
	at org.apache.cxf.jaxrs.client.SyncInvokerImpl.method(SyncInvokerImpl.java:135)
	at org.apache.cxf.jaxrs.client.SyncInvokerImpl.method(SyncInvokerImpl.java:130)
	at org.apache.cxf.jaxrs.client.SyncInvokerImpl.get(SyncInvokerImpl.java:50)
	at org.apache.cxf.jaxrs.client.spec.InvocationBuilderImpl.get(InvocationBuilderImpl.java:88)
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.queryVehicles(TeslaAccountHandler.java:209)
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.lambda$1(TeslaAccountHandler.java:155)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.IOException: IOException invoking https://owner-api.teslamotors.com/api/1/vehicles: HTTPS hostname wrong:  should be <owner-api.teslamotors.com>
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.mapException(HTTPConduit.java:1402)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1386)
	at org.apache.cxf.io.AbstractWrappedOutputStream.close(AbstractWrappedOutputStream.java:77)
	at org.apache.cxf.transport.AbstractConduit.close(AbstractConduit.java:56)
	at org.apache.cxf.transport.http.HTTPConduit.close(HTTPConduit.java:673)
	at org.apache.cxf.interceptor.MessageSenderInterceptor$MessageSenderEndingInterceptor.handleMessage(MessageSenderInterceptor.java:63)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.jaxrs.client.AbstractClient.doRunInterceptorChain(AbstractClient.java:705)
	at org.apache.cxf.jaxrs.client.WebClient.doChainedInvocation(WebClient.java:1081)
	... 15 common frames omitted
Caused by: java.io.IOException: HTTPS hostname wrong:  should be <owner-api.teslamotors.com>
	at java.base/sun.net.www.protocol.https.HttpsClient.checkURLSpoofing(HttpsClient.java:648)
	at java.base/sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:581)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1581)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1509)
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:527)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:329)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream$2.run(URLConnectionHTTPConduit.java:377)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream$2.run(URLConnectionHTTPConduit.java:373)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.getResponseCode(URLConnectionHTTPConduit.java:373)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.doProcessResponseCode(HTTPConduit.java:1599)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponseInternal(HTTPConduit.java:1627)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponse(HTTPConduit.java:1572)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1373)
	... 22 common frames omitted
```

**Comment by cguedel** [comment]
Interesting... could be that you get throttled by Tesla's WAF and the logon page now looks different. Will need to see if I can reproduce that.

The second thing looks very odd and not really related to the changes, more like a certificate issue?

**Comment by kaikreuzer** [comment]
Will try to investigate myself as well. FTR: I am running the binding from within the IDE.

**Comment by cguedel** [comment]
I was doing the same, Eclipse on Windows with Java 11, FWIW.

**Comment by cguedel** [comment]
I know there is still one problem with the SSO handler. In case you successfully logged in on the SSO endpoint using username/password and try this again, it directly skips the sign in step upon fetching the login form. This is currently not (yet) handled by the handler.

But never saw those two issues/exceptions you just posted

**Comment by kaikreuzer** [comment]
The exceptions are not fully reproducible. Now the discovery starts and I can see in the debugger that it has the correct vehicleArray, but the call in [this line](https://github.com/openhab/openhab-addons/pull/10259/files#diff-7ac9af9aea4bd8cf22da6ddde6ee480b06287b656312f1862f72410c5d362b45L395) results in an HTTP 408, which again makes the discovery return no results. Pretty weird...

**Comment by cguedel** [comment]
Ah, the 408 means that the vehicle is asleep and you need to wake it up first. That's a trap that was in there before and I also fell into it the first time the discovery of the account actually worked :)

**Comment by kaikreuzer** [comment]
Thanks - indeed, waking up the vehicle made it being discovered. Maybe something that could be improved, but not within this PR.
So good news is that account+vehicle things are now online. Will see if it stays like that and will meanwhile do the code review!

**Comment by kaikreuzer** [comment]
Another exception that asks to be fixed:
```
19:18:47.000 [-thingHandler-3] ERROR o.o.b.t.i.h.TeslaAccountHandler:472 - An exception occurred while executing a request to the vehicle: 'null'
java.lang.NullPointerException: null
	at org.openhab.binding.tesla.internal.handler.TeslaVehicleHandler.getVehicleId(TeslaVehicleHandler.java:227)
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler$Request.run(TeslaAccountHandler.java:466)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.openhab.binding.tesla.internal.throttler.QueueChannelThrottler.lambda$0(QueueChannelThrottler.java:45)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeRequest.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.openhab.binding.tesla.internal.protocol.SSO;
+
+import static org.openhab.binding.tesla.internal.TeslaBindingConstants.*;
+
+/**
+ * The {@link AuthorizationCodeExchangeRequest} is a datastructure to exchange
+ * the authorization code for an access token on the SSO endpoint
+ *
+ * @author Christian Güdel
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeRequest.java:21]:
```suggestion
 * @author Christian Güdel - Initial contribution
```

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeResponse.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tesla.internal.protocol.SSO;
+
+/**
+ * The {@link AuthorizationCodeExchangeResponse} is a datastructure to capture
+ * the response of an {@link AuthorizationCodeExchangeRequest}
+ *
+ * @author Christian Güdel
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeResponse.java:19]:
```suggestion
 * @author Christian Güdel - Initial contribution
```

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/RefreshTokenRequest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package org.openhab.binding.tesla.internal.protocol.SSO;
+
+import static org.openhab.binding.tesla.internal.TeslaBindingConstants.*;
+
+/**
+ * The {@link RefreshTokenRequest} is a datastructure to refresh
+ * the access token for the SSO endpoint
+ *
+ * @author Christian Güdel
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/RefreshTokenRequest.java:21]:
```suggestion
 * @author Christian Güdel - Initial contribution
```

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeResponse.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tesla.internal.protocol.SSO;
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeResponse.java:13]: Please stick to lower case package names.
```suggestion
package org.openhab.binding.tesla.internal.protocol.sso;
```

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/TokenExchangeRequest.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tesla.internal.protocol.SSO;
+
+/**
+ * The {@link TokenExchangeRequest} is a datastructure to exchange
+ * the access token from the SSO endpoint for an owners API access token
+ *
+ * @author Christian Güdel
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/TokenExchangeRequest.java:19]:
```suggestion
 * @author Christian Güdel - Initial contribution
```

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Code wise it looks excellent! Looking forward to more contributions from you ;-)
A few small change requests below.

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+import java.util.Iterator;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.lang.RandomStringUtils;
```
[the remark this answers is not in the document]
Comment (cguedel) [reply bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java]: Oh sorry, did not see that. Of course, now replaced with pure java

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+    private ContentResponse executeHttpRequest(org.eclipse.jetty.client.api.Request request) {
+        request.timeout(10, TimeUnit.SECONDS);
+
+        ContentResponse response;
+        try {
+            response = request.send();
+            return response;
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            logger.error("An exception occurred while invoking a HTTP request: '{}'", e.getMessage());
```
[the remark this answers is not in the document]
Comment (cguedel) [reply bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java]: Done, most of the logging changed to debug. If the SSO login or token renewal return null, the binding will go to OFFLINE stating a configuration error. I need this so the handler doesn't retry over and over again to login and then hit the WAF of Tesla (stated in the API docs).

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link TeslaSSOHandler} is responsible for authenticating with the Tesla SSO service.
+ *
+ * @author Christian Güdel - Initial contribution
+ */
+public class TeslaSSOHandler {
```
[the remark this answers is not in the document]
Comment (cguedel) [reply bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java:55]: Done

File: bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeRequest.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ */
+@SuppressWarnings("unused")
```
[the remark this answers is not in the document]
Comment (cguedel) [reply bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/protocol/SSO/AuthorizationCodeExchangeRequest.java]: Done

**Comment by cguedel** [comment]
@kaikreuzer Thanks for the quick review! I added the changes requested. Not sure what to do about the exceptions though. The very first one with the buffer capacity error seems very weird. The others I would say were already in :)

**Comment by kaikreuzer** [comment]
> The very first one with the buffer capacity error seems very weird.

I haven't seen it ever again, so let's ignore it.
The "HTTPS hostname wrong" occurs quite regular for me, though, and it makes the vehicle Thing go OFFLINE for a while.
I traced it down to `HttpsClient.checkURLSpoofing()`, where there are no peer certificates available and the session is marked as invalidated. No idea how this happens, though.

I am fine to merge it like it is, though, and we can see what testing by others bring up. I might do a follow-up PR for handling this exception gracefully then.

**Review by kaikreuzer — APPROVED** [review]
Many thanks, let's get that into the SNAPSHOT distro and have people test it!

**Comment by kaikreuzer** [comment]
FTR: The fix is now included in distro build 2230 - would be great if everyone could test it!

**Comment by kaikreuzer** [comment]
@cguedel For the hostname exception, I have created https://github.com/openhab/openhab-addons/issues/10264 for further tracking. It also occurs on my production system, so it wasn't due to IDE or something like that.

Checking my log file, it occurs as often as every 4 minutes, so I guess we'll need to do something about it...

**Comment by phuongpham** [comment]
Just downloaded the lastest snapshot from  https://openhab.jfrog.io/artifactory/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.tesla/3.1.0-SNAPSHOT/org.openhab.binding.tesla-3.1.0-SNAPSHOT.jar

Everything works fine!
@cguedel and @kaikreuzer  great work!

**Comment by kaikreuzer** [comment]
Cherry-picked it to `3.0.x`.

**Comment by mrphil81** [comment]
@cguedel and @kaikreuzer Thank you very much for your effort! I installed the snapshot jar file on my openhab version 3.0.1. Works like a charm again since 2 days. I also recognized the 408 exception (timeout due to sleeping vehicle), but it wakes up within a short time. I use it mainly to start/stop charging and display some vehicle information.
Thanks again, highly appreciated!

**Comment by ScanxTaz** [comment]
Stupid question, but is there any update on this binding ?

**Pull request merged**